[MAX] Propose to Add Support for Visual Generation: `Tongyi-MAI/Z-Image`

tolgacangoz opened the pull request:
Hey Modular community!

This PR is in the process of introducing a dedicated pipeline for diffusion-based vision models, specifically implementing the Z-Image architecture. While MAX currently has strong support for LLMs and VLMs, support for multi-part diffusion systems (VAE(s) + Text Encoder(s) + (maybe Image Encoder) + (maybe Audio Encoder) + Backbone Transformer(s) + Scheduler/Sampler/Solver) is an area for growth opportunity.

Inspired by the Democratizing AI Compute series, I aim to demonstrate how MAX can optimize multi-part, multi-modal graphs without relying heavily on lossy approximations like distillation or caching intermediate latents, etc.

I encountered benchmarks for possible diffusion models by testing the U-Net on the website, which was the earlier denoiser for diffusion-based generative vision models. I propose the natural next steps.

<details>
<summary><strong>⚡️- Image: An Efficient Image Generation Foundation Model with Single-Stream Diffusion Transformer</strong></summary>

- The current text-to-image generation SOTA* in the open-source/weights category: https://artificialanalysis.ai/image/leaderboard/text-to-image?open-weights=true
- GitHub repo: https://github.com/Tongyi-MAI/Z-Image
- Z-Image gallery: https://modelscope.cn/studios/Tongyi-MAI/Z-Image-Gallery

I am using the pipeline design from Hugging Face's `diffusers`.


*: 29.12.2025. It seems that there are now two models that have already surpassed Z-Image-Turbo without even merging this PR; i.e., `fal/FLUX.2-dev-Turbo` with FLUX [dev] Non-Commercial License v2.0 and `Qwen/Qwen-Image-2512` with Apache-2.0. The second one seems more appropriate in terms of commercial usage, but it is `~3` times larger than Z-Image. `fal/FLUX.2-dev-Turbo` is much larger than Z-Image. Thus, I think, Z-Image is still the most intelligence/value-compressed model.

</details>

TODOs:
<details>
<summary><strong> ✅ :question: 1. Make it work</strong></summary>

- :question: The VAE's decoder and the backbone transformer seem to be compiled, but the text encoder's -`Qwen/Qwen3-4B`- compilation seems stuck without any warning. I have tried to fix it, but couldn't manage, and since I am not familiar with compiling internals, I opened an issue.
- The overall pipeline seems to "work" (except the text encoder), but generates `NaN`s. I am working on it at the second stage.

</details>

<details>
<summary><strong>✅ :hourglass_flowing_sand: 2. Make it right</strong></summary>

- I fixed the `NaN` issue of the first stage. I accidentally coded an equation wrongly :sweat_smile:.

<details>
<summary>✅ :grey_question: 1. Make it right functionally</summary>

The comparison of the outputs between `diffusers` and this PR with the same inputs, including the `seed==0`:
```txt
prompt="A high-quality 3D render of joyful Pikachu standing in a lush nature environment,
		holding a large white sign board with the text "Image Generation is awesome in
		Modular!" written on it. A neon lightning-bolt lamp (⚡️) is floating in the blue
		sky. Bright lighting, vibrant colors, sharp focus, cinematic composition."
height=1024
width=1024
num_inference_steps=9	# This actually results in 8 DiT forwards
guidance_scale=0.0		# Guidance should be 0 for the Turbo models
generator=torch.Generator("cpu").manual_seed(0)
```

<details>
<summary>Previous</summary>

<table>
<tr>
 <td><code>diffusers</code>
 <td>This PR's Eager with Lazy Evaluation
 <td>This PR's Compiled
<tr>
 <td><img width="230" height="230" alt="diffusers" src="https://github.com/user-attachments/assets/d64d4750-b743-4eb2-bd61-87614946d1d3" />
 <td><img width="230" height="230" alt="eager" src="https://github.com/user-attachments/assets/2395029c-19c4-438f-ad7e-9bc7e842c76d" />
 <td><img width="230" height="230" alt="compiled" src="https://github.com/user-attachments/assets/58d34794-1136-430e-87ae-87a4e443d3fe" />
</table>

I am investigating the issue for the compiled one's output.

</details>

<details>
<summary>Now</summary>

<table>
<tr>
 <td><code>diffusers</code>
 <td>This PR's Eager with Lazy Evaluation
 <td>This PR's Compiled
<tr>
 <td><img width="230" height="230" alt="diffusers" src="https://github.com/user-attachments/assets/d64d4750-b743-4eb2-bd61-87614946d1d3" />
 <td><img width="230" height="230" alt="eager" src="https://github.com/user-attachments/assets/2395029c-19c4-438f-ad7e-9bc7e842c76d" />
 <td><img width="230" height="230" alt="compiled2" src="https://github.com/user-attachments/assets/1fc07db0-3703-475c-9dbe-10d839c6a433" />
</table>

When I remove weight loading from `.compile()`, and apply `load_state_dict()` before compiling, it works. Maybe there is an issue with loading and compiling in the same function:
https://github.com/modular/modular/blob/c168077e49ef782f5e05d4522dec67a81f460135/max/python/max/nn/module_v3/module.py#L504-L508

</details>

- Since I couldn't run the text encoder -`Qwen/Qwen3-4B`- properly, I generated `prompt_embeds` by using PyTorch, then loaded it manually into the pipeline.
- I generated the first noise tensor with `torch.Generator("cpu").manual_seed(0)` to equalize the random number generators.
- :grey_question: I didn't set the GPU-related environment variables for 100% determinism yet. I will evaluate why the two images are not 100% the same.
- The compilation process takes `~11` minutes, and the first-time full forward pass of the pipeline takes `~4` minutes. I only compile the VAE's decoder and the transformer.
- I compiled the whole pipeline's `execute()` by including the whole denoising loop and the VAE's decoding. The first inference takes `~60` seconds.
- Testing **superficially** with a glance: Second usage of the pipeline takes `~30` seconds in both cases. Seems `10+` times slower than `diffusers` in terms of latency. I am investigating this in the third stage.

</details>

<details>
<summary>:hourglass_flowing_sand: 2. Make it right in terms of <code>modular</code>'s pipeline and serving design</summary>

- I will add potential tests (templates) too.

</details>

</details>

<details>
<summary>:soon: Example code for <code>Tongyi-MAI/Z-Image-Turbo</code> with MAX Serve.</summary>

```sh
uv pip install grpcio -qU
uv pip install modular --index-url https://dl.modular.com/public/nightly/python/simple --prerelease allow -q

MODULAR_PY_SITE=$(python3 - <<'EOF'
import site
print(site.getsitepackages()[0])
EOF
)
export LD_LIBRARY_PATH="$MODULAR_PY_SITE/modular/lib:${LD_LIBRARY_PATH:-}"
export PYTHONPATH="$HOME/modular/max/python:${PYTHONPATH:-}"
export MODULAR_MAX_DEBUG="True"

git clone --depth 1 --single-branch --branch <TBA> https://github.com/tolgacangoz/modular.git && cd modular/max/python

max serve --model Tongyi-MAI/Z-Image-Turbo \
--custom-architectures max/pipelines/architectures/z_image_module_v3 \
--task pixel_generation
```

```sh
curl -X POST http://localhost:8000/v1/images/generations \
  -H "Content-Type: application/json" \
  -d '{
    "messages": [{
                  "role": "user",
                  "content": "A cute baby sea otter floating in crystal clear water"
                  }],
    "chat_template_options": [{
                  "add_generation_prompt": True,
                  "enable_thinking": True
                  }],
    "n": 1,
    "size": "1024x1024",
    "guidance_scale": 0.0,
    "num_inference_steps": 9
  }' -o result.json

python -c "import json,base64;d=json.load(open('generated_image.json'));open('generated_image.png','wb').write(base64.b64decode(d['data'][0]['b64_json']))"
```

</details>

<details>
<summary><strong>:hourglass_flowing_sand: 3. Make it fast</strong></summary>

1. Make it faster than `diffusers`; i.e., PyTorch's Eager mode.

2. Make it faster than `diffusers`' [`flux-fast`](https://github.com/huggingface/flux-fast) version of Z-Image-Turbo; i.e., `torch.compile(...)` etc...

</details>

<details>

<summary><strong> :pause_button: 4. Make it the fastest</strong></summary>

**NOTE: Since this stage includes a relatively long study plan, I am postponing it. For this PR, it might be proper to merge it when the third stage is completed. Contribution of kernels would be a completely separate PR anyway.**

1. I plan to begin solving Mojo 🔥 GPU Puzzles after the completion of reading several GPU optimization-related posts, such as the Matrix Multiplication on Blackwell blog post series.
2. Studying and implementing throughput efficiency strategies.
3. Planned benchmarks for throughput efficiency:

<table>
<tr>
 <td><img width="13" height="13" alt="150859076" src="https://github.com/user-attachments/assets/8cd2253b-ff7f-46fd-8a8b-d4aabf808988" /> <code><b>modular</b></code>
 <td><img width="13" height="13" alt="136984999" src="https://github.com/user-attachments/assets/943ac84e-a8cc-42a9-a818-81cd83b7c82c" /> <code><b>vllm-omni</b></code>
 <td><img width="13" height="13" alt="147780389" src="https://github.com/user-attachments/assets/cb6f9cce-671b-411e-864f-0a3a7da5836e" /> <code><b>sglang</b></code>
 <td><img width="13" height="13" alt="147780389" src="https://github.com/user-attachments/assets/4fa1ff1e-7341-44ad-95f1-f32fa58ac2fd" /> <code><b>stable-diffusion.cpp</b></code>
 <td><img width="13" height="13" alt="25720743" src="https://github.com/user-attachments/assets/f7b89138-25ed-4e76-b4fd-233f625ed7df" /> <code><b>diffusers</b></code>
<tr>
 <td>
 <td>
 <td>
 <td>
 <td>
<tr>
 <td>
 <td>
 <td>
 <td>
 <td>
</table>

</details>

KCaverly commented on max/python/max/pipelines/architectures/z_image_module_v3/arch.py:
```diff
@@ -0,0 +1,57 @@
+    scheduler=FlowMatchEulerDiscreteScheduler,
+    vae=AutoencoderKL,
+    text_encoder=qwen3_arch,
+    tokenizer=TextTokenizer,
```
Are you looking to implement a custom tokenizer with this model? The TextTokenizer will not be able to produce a `ZImageTextAndVisionContext`.

tolgacangoz replied:
`TextTokenizer` is seen as a tokenizer of Qwen3, which is the text encoder of the pipeline; thus, I added it.
I just removed `ZImageTextAndVisionContext`, since this pipeline is supposed to accept text-only inputs.

KCaverly replied:
Awesome, that makes this a bit simpler. If `TextTokenizer` + `TextContext` is not enough, let me know and I am happy to take a look and help.

tolgacangoz replied:
I wasn't knowledgeable enough one month ago, but now it turns out that `TextTokenizer.next_context()` seems to return text-generation-specialized `TextContext`, [thus I proposed to modify `TextTokenizer`](https://github.com/modular/modular/pull/5604/changes#diff-8dafd283fe1be08c9c5b19dfbe5bd5c46e9de327d853ef1353aff7d6c404ac9a).

KCaverly commented on max/python/max/pipelines/architectures/z_image/model.py:
```diff
@@ -0,0 +1,1034 @@
+
+            # Extract parameters from extra_model_args or use defaults
+            # TextAndVisionContext extra_model_args values are numpy arrays.
+            width = int(ctx.extra_model_args.get("width", np.array([1024])).item()) if "width" in ctx.extra_model_args else 1024
```
Non-Blocking: Since extra_model_args is a dict, you can do:
```
int(ctx.extra_model_args.get("width", 1024))
```

Which will return you the 1024 default, when 'width' is not provided in the dict.

tolgacangoz replied:
This function is there as a template for now, don't see it :sweat_smile:.

KCaverly commented on max/python/max/pipelines/architectures/z_image/model.py:
```diff
@@ -0,0 +1,1034 @@
+    max_sequence_length: int = 512
+    """ The maximum sequence length for the input sequence."""
+
+    #signal_buffers: list[Tensor]
```
Having this commented out here, contradicts support for multi_gpu listed in the SupportedArchitecture. I would suggest, for initial bringup you drop support for multi-gpu. Will likely make this initial work a bit easier.

tolgacangoz replied:
There are still such commented-out, relic parts since the PR is still in draft.
I agree with removing multi-GPU support for now.

KCaverly reviewed (commented):
Awesome progress! Just added a few drive by comments.

KCaverly commented on max/python/max/pipelines/lib/image_generator_pipeline.py:
```diff
@@ -0,0 +1,92 @@
+            sum(ctx.active_length for ctx in inputs.batch.values())
+        )
+
+        next_chunk = getattr(self.pipeline_model, "next_chunk")  # type: ignore[has-type]  # noqa: B009
```
Awesome work stubbing this out! I think we should be able to use two APIs to get this to work:

```
# Retrieve the inputs
model_inputs = self.pipeline_model.prepare_initial_token_inputs(inputs.batches)

# Execute the model
model_outputs = self.pipeline_model.execute(model_inputs)
```

One other thought, is that the model_outputs is expected to be different than the ImageGenerationOutput. Commonly pipeline models will only return logits etc, and then the pipeline itself will do post-processing as well, to return a ImageGenerationOutput.

tolgacangoz replied:
I haven't been able to focus on the PR lately. I have another commitment, sorry; I plan to get back by Thursday.

tolgacangoz committed:
feat: Add pixel generation pipeline support

Introduces `PixelGenerationPipeline` and `PixelGenerationConfig` to enable
image/video generation capabilities alongside existing text and audio tasks.

Adds `PIXEL_GENERATION` task type with corresponding context, input, output,
and metadata types to the interfaces layer.

Registers new `z_image_module_v3` architecture and integrates `diffusers`
configuration support for pixel generation workflows.

Updates CLI serve and API server to route pixel generation requests to the
appropriate pipeline implementation.

tolgacangoz committed:
Add `PixelGenerationConfig` for image/video pipelines

Introduces configuration infrastructure for diffusers-style image and video
generation pipelines.

Adds `PixelGenerationType` enum to specify generation modes (text-to-image,
text-to-video, image-to-image, inpainting, controlnet, etc.) with properties
to determine input/output requirements.

Adds `DiffusersConfig` and `DiffusersComponentConfig` classes to parse
multi-folder diffusers repository structures from local paths or HuggingFace,
extracting component metadata, configs, and weight paths from model_index.json.

Adds `PixelGenerationConfig` extending `PipelineConfig` to support these
pipelines with server-side settings like generation type and scheduler
override, while inference parameters come from API requests.

tolgacangoz committed:
feat: Add support for returning second-to-last hidden state in Transformer

Add `SECOND_TO_LAST` option for hidden state returns

tolgacangoz committed:
Add `ZImagePipeline` architecture for image generation

Introduces a new `SupportedArchitecture` configuration for the Z-Image
pipeline, enabling pixel generation capabilities.

Integrates transformer, VAE, scheduler, and text encoder components
to support the Tongyi-MAI/Z-Image-Turbo model.

tolgacangoz committed:
Add model configuration for ZImage image generation

Introduces `ZImageConfig` and supporting config classes for the ZImage
diffusion model architecture, including `SchedulerConfig`, `VAEConfig`,
and `TransformerConfig`.

Delegates text encoder handling to `Qwen3Config` with prefix caching
disabled since diffusion models don't use autoregressive KV caching.
Uses second-to-last hidden states for conditioning to match diffusers
behavior.

tolgacangoz committed:
Add Z-Image text-to-image model architecture

Introduces `ZImage` container class and `ZImageModel` pipeline for
text-to-image generation using a diffusion-based architecture.

The implementation includes:
- `ZImageInputs` dataclass for flexible input configuration
- Flow-matching Euler discrete scheduler integration
- VAE decoder for latent-to-image conversion
- Transformer backbone for denoising steps
- Support for classifier-free guidance and callback hooks

Components are compiled separately for optimal performance, with
weights partitioned by their source directory (vae/, text_encoder/,
transformer/) to support diffusers-style model layouts.

tolgacangoz committed:
Add native `FlowMatchEulerDiscreteScheduler` for diffusion

Implements a native Modular port of the `FlowMatchEulerDiscreteScheduler`
from diffusers for use in visual generation pipelines.

Supports multiple sigma schedule types including Karras, exponential,
and beta distributions, as well as dynamic resolution-dependent
timestep shifting with exponential or linear modes.

Pre-computes sigmas and timesteps as Python lists to avoid device
synchronization overhead during inference steps.

tolgacangoz committed:
Add VAE neural network modules for visual generation

Introduces a complete set of neural network components for Variational
Autoencoder (VAE) with KL divergence loss, enabling image encoding and
decoding in latent diffusion pipelines.

Key components added:
- `AutoencoderKL` with encoder/decoder architecture and tiling support
- `Encoder` and `Decoder` classes with configurable block types
- `UNetMidBlock2D`, `DownEncoderBlock2D`, `UpDecoderBlock2D` for UNet structure
- `ResnetBlock2D` for residual connections
- `Upsample2D` and `Downsample2D` for spatial resolution changes
- `DiagonalGaussianDistribution` for latent space sampling
- Custom layers including `Conv2d`, `GroupNorm`, `LayerNorm`, `Attention`

Supports memory-efficient tiled encoding/decoding for high-resolution
images and sliced batch processing for large batch sizes.

tolgacangoz committed:
Add `ZImageTransformer2DModel` for image generation

Introduces a graph-compilable transformer architecture for image
generation with support for:

- Single-stream attention with RoPE embeddings
- AdaLN modulation for timestep conditioning
- Patchification and unpatchification of image latents
- Caption feature processing with context refiners
- Flash attention GPU kernels for efficient computation

The implementation includes `TimestepEmbedder`, `RopeEmbedder`,
`ZImageTransformerBlock`, and `FinalLayer` components designed
for batch_size=1 inference with fixed shape compilation.

tolgacangoz committed:
Add `diffusers` model support to tokenizer and config

Extends `MAXModelConfig` with `DiffusersConfig` integration for handling
diffusers-style models that use `model_index.json` configuration files.

Updates `TextTokenizer` to support `PixelGenerationRequest` and create
`PixelContext` objects for image generation workflows, including handling
of negative prompts and diffusion-specific parameters.

Adds subfolder support for loading tokenizers from nested model directories.

tolgacangoz committed:
Add pixel generation pipeline support for diffusers

Extends the pipeline registry to support `PixelGenerationPipeline` and
`PixelGenerationContext` for diffuser-based image generation models.

Updates config handling to properly retrieve text encoder config from
`_diffusers_config` instead of the main model config for diffusers models.

Adds proper type handling for `PixelGenerationRequest` in tokenizer casting
and context type retrieval for diffuser model pipelines.

tolgacangoz committed:
Add pixel generation pipeline for diffusion models

Introduces infrastructure for image generation using diffusion models:

- Adds `PixelGenerationRequest`, `PixelGenerationContext`, and
  `PixelGenerationOutput` interfaces for handling image generation
- Creates `PixelGenerationPipeline` to execute diffusion-based inference
- Implements `PixelGeneratorPipeline` serving layer with image postprocessing
- Extends OpenAI routes with `/images/generations` endpoint
- Adds `PixelContext` for managing image generation request state

This enables MAX to serve image generation models through a standardized
API compatible with OpenAI's image generation interface.

tolgacangoz committed:
Add initial pixel generation scheduler utilities for metrics

tolgacangoz committed:
Improves safetensors dtype detection just for proposal purpose

tolgacangoz committed:
`pre-commit run --all-files`

tolgacangoz commented on max/python/max/interfaces/task.py:
```diff
@@ -34,6 +34,7 @@
     - EMBEDDINGS_GENERATION: Generate vector embeddings for input data.
     - AUDIO_GENERATION: Generate audio outputs from input data.
     - SPEECH_TOKEN_GENERATION: Generate speech tokens for speech-related tasks.
+    - PIXEL_GENERATION: Generate/Edit images/videos from input data.
```
Which one is better: `PIXEL_GENERATION` or `VISUAL_GENERATION`, or another alternative?

KCaverly replied:
PIXEL_GENERATION is fine for now, fairly easy to rename if we need to rework it later.

tolgacangoz commented on max/python/max/pipelines/architectures/z_image_module_v3/model.py:
```diff
@@ -0,0 +1,944 @@
+    def progress_bar(self, total: int) -> tqdm:
+        return tqdm(total=total)
+
+    def execute(
```
I imitated `diffusers`' pipeline's `__call__` method here, should I convert it to something like that:

```py
def execute(self, model_inputs: ModelInputs) -> ModelOutputs:
    """Execute the diffusion pipeline."""
    model_inputs = cast(ZImageInputs, model_inputs)

    # Run diffusion denoising loop
    latents = self._run_denoising(model_inputs)

    # Decode with VAE
    image = self._decode_latents(latents)

    return ModelOutputs(image=image)
```

tolgacangoz commented on max/python/max/nn/transformer/transformer.py:
```diff
@@ -86,6 +86,7 @@ class ReturnHiddenStates(str, Enum):
     ALL = "all"
     LAST_NORMALIZED = "last_normalized"
     ALL_NORMALIZED = "all_normalized"
+    SECOND_TO_LAST = "second_to_last"
```
What would be the best way to use this option?

```py
prompt_embeds = self.text_encoder(
    input_ids=text_input_ids,
    attention_mask=prompt_masks,
    output_hidden_states=True,
).hidden_states[-2]
```

tolgacangoz replied:
This doesn't seem proper. We need an encoder version of it; that is, `Qwen3Model` instead of `Qwen3ForCausalLM`.

tolgacangoz commented on max/python/max/pipelines/lib/diffusers_config.py:
```diff
@@ -0,0 +1,354 @@
+at the root that maps component names to their classes and subfolders. This
+module provides utilities for parsing and representing this structure.
+
+Example structure:
```
What would be the best way to add usability to `diffusers`' repos?

tolgacangoz committed:
fix: Fix several bugs and typos

tolgacangoz committed:
feat: Replace `F.unsqueeze` and `F.squeeze` with their corresponding tensor methods.

tolgacangoz committed:
Refactor `z_image_module_v3` with lazy device moves

Wraps device transfers in `F.lazy()` context for transformer and VAE
components to improve initialization behavior.

Replaces deprecated `F.range` calls with `F.arange` across multiple
modules for consistency.

Comments out unused architecture component registrations (`scheduler`,
`vae`, `text_encoder`, `transformer`) and removes their imports to reduce
dependencies.

Reorders `dataclass` fields in `DiffusersConfig` to fix field ordering
issues with default values.

tolgacangoz committed:
fix: Add optional pixel/visual generation parameters

tolgacangoz committed:
fix: Refactor the `PixelContext` class to better suit visual generation pipelines.

tolgacangoz committed:
fix: Improve support for `diffusers` models

tolgacangoz committed:
update: Refactors `ZImageConfig` inheritance

Simplifies the `ZImageConfig` inheritance structure by removing the intermediate base class and directly inheriting from the `MAXModelConfigBase`.

Also, removes unused copyright notice in `resnet.py`.

tolgacangoz commented on max/python/max/pipelines/architectures/z_image_module_v3/nn/resnet.py:
```diff
@@ -0,0 +1,252 @@
+# ===----------------------------------------------------------------------=== #
+# Copyright (c) 2025, Modular Inc. All rights reserved.
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions:
+# https://llvm.org/LICENSE.txt
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ===----------------------------------------------------------------------=== #
+
+# Copyright 2025 The HuggingFace Team. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
```
Do we have to include "The HuggingFace Team. All rights reserved." too, since I am using their files as general templates? But, the Apache License v2.0 with LLVM Exceptions is supposed to include HF's the Apache License v2.0, right 🤔.

KCaverly replied:
@BradLarson I dont think we have to do this right?

tolgacangoz commented on max/python/max/pipelines/lib/hf_utils.py:
```diff
@@ -353,6 +356,7 @@ def __hash__(self) -> int:
                 self.repo_id,
                 self.revision,
                 self.trust_remote_code,
+                self.subfolder,
                 self.repo_type,
             )
         )
```
Since `subfolder` has been `None` by default, this addition would be backward compatible, right?

tolgacangoz committed:
fix typos

tolgacangoz committed:
fix: Refactor pixel generation context and inputs

Refactors the pixel generation context to a protocol, `PixelGenerationContext`, defining the required properties for diffusion model pipelines.

Updates `PixelGenerationInputs` to support a variable list of batches, allowing for more flexible handling of requests.  This change improves the structure and type safety of pixel generation pipelines.

tolgacangoz committed:
fix: Rename `model_name` to `model`

tolgacangoz committed:
fix/update: Replace PyTorch random generation with native Modular API

Removes PyTorch dependency for random tensor generation in the image
pipeline by using `random.gaussian` and `set_seed` from the native
Modular experimental API.

Adds `_SimpleProgressBar` class to replace `tqdm` dependency temporarily
until it's added to project dependencies.

Comments out unused code for loading pre-computed prompt embeddings
and preserves PyTorch-based generation as comments for `diffusers` parity
testing when needed.

KCaverly commented on max/python/max/interfaces/pipeline_variants/pixel_generation.py:
```diff
@@ -0,0 +1,315 @@
+    """
+    Negative prompt to guide what NOT to generate.
+    """
+    messages: list[TextGenerationRequestMessage] | None = None
```
If this is not used downstream, lets remove it from here.

tolgacangoz replied:
https://github.com/modular/modular/pull/5604/changes#r2685626093

KCaverly commented on max/python/max/interfaces/pipeline_variants/pixel_generation.py:
```diff
@@ -0,0 +1,315 @@
+    responses.
+    """
+
+    def __post_init__(self) -> None:
```
Further to the removal of messages, if messages is removed, so should this post init validation.

tolgacangoz replied:
Using whether `prompt` or `messages` will be decided after this is resolved: https://github.com/modular/modular/pull/5604/changes#r2683986177

KCaverly commented on max/python/max/interfaces/pipeline_variants/pixel_generation.py:
```diff
@@ -0,0 +1,315 @@
+    completion APIs, where each message represents a turn in the conversation.
+    If provided, the model will generate responses based on these messages.
+    """
+    image: bytes | None = None
```
Alternatively, it does not appear that this is currently used in your downstream pipeline. Please remove, we can add these in as new models require them.

tolgacangoz replied:
I wanted to add other tasks too as a placeholder for now, but the addition of them seems to break the atomicity of the PR, thus removing them all.

KCaverly commented on max/python/max/interfaces/pipeline_variants/pixel_generation.py:
```diff
@@ -0,0 +1,315 @@
+    """
+    The conditioning image to use for image-to-image/video generation.
+    """
+    video: bytes | None = None
```
Similar to image above, if we are not using this field, lets not add it prematurely.

tolgacangoz replied:
[Removing.](https://github.com/modular/modular/pull/5604/changes#r2685636474)

KCaverly commented on max/python/max/interfaces/pipeline_variants/pixel_generation.py:
```diff
@@ -0,0 +1,315 @@
+    """
+    The conditioning video for video-to-video generation.
+    """
+    mask: bytes | None = None
```
Ditto to the comment on video/image above.

tolgacangoz replied:
[Removing.](https://github.com/modular/modular/pull/5604/changes#r2685636474)

KCaverly commented on max/python/max/interfaces/pipeline_variants/pixel_generation.py:
```diff
@@ -0,0 +1,315 @@
+    """
+    Width of generated image/frame in pixels. Defaults to model's default (typically 1024).
+    """
+    num_frames: int = 1
```
Since we are currently supporting just image generation today, please remove this until we have a functional video generation pipeline.

tolgacangoz replied:
[Removing.](https://github.com/modular/modular/pull/5604/changes#r2685636474)

KCaverly commented on max/python/max/interfaces/pipeline_variants/pixel_generation.py:
```diff
@@ -0,0 +1,315 @@
+    """
+    Number of frames for video generation. 1 = single image, >1 = video.
+    """
+    fps: int = 24
```
Ditto to the num_frames comment above.

tolgacangoz replied:
[Removing.](https://github.com/modular/modular/pull/5604/changes#r2685636474)

KCaverly commented on max/python/max/interfaces/pipeline_variants/pixel_generation.py:
```diff
@@ -0,0 +1,315 @@
+    """
+    Number of images/videos to generate per prompt.
+    """
+    chat_template_options: dict[str, Any] | None = None
```
This appears to not be used downstream, lets remove it completely for now.

tolgacangoz replied:
I added this to make the least modification at `TextTokenizer.new_context()`, since it has:
https://github.com/modular/modular/blob/38c1a3745c5fdffacc816ae14a6d47bef8660f40/max/python/max/pipelines/lib/tokenizer.py#L489-L497

And to be able to use them: https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/z_image/pipeline_z_image.py#L217-L222

[I decided to use and modify the current `TextTokenizer` rather than creating a new one.](https://github.com/modular/modular/pull/5604#discussion_r2679604673) Because `TextTokenizer` already has what we would need for visual-gen, too. It just has more features.

KCaverly replied:
Given how different the intention of the code is, and the paths in which they are used, I think we should create a new PipelineTokenizer for PixelGeneration. I am a bit weary about introducing PixelGeneration related code into our TextTokenizer, as that class is used as the base for all tokenizers currently in production, introducing a bunch of Pixel related functionality where it is not used. 

Additionally, typed unions in this way, have caused headaches in the past with type checking downstream. I would suggest, we separate this code out into a new PixelTokenizer. Despite the duplication, there are likely other ways we can remove duplicated code in the future and refactor as needed.

Introducing Tokenizer + tests, seems like a pretty logical second PR we can break this into and merge after Interfaces.

KCaverly commented on max/python/max/interfaces/pipeline_variants/pixel_generation.py:
```diff
@@ -0,0 +1,315 @@
+    """
+    Optional dictionary of options to pass when applying the chat template.
+    """
+    tools: list[TextGenerationRequestTool] | None = None
```
Until tool use is supported, please remove.

tolgacangoz replied:
I added this to make the least modification at `TextTokenizer.new_context()`, since it uses `tools`:
https://github.com/modular/modular/blob/38c1a3745c5fdffacc816ae14a6d47bef8660f40/max/python/max/pipelines/lib/tokenizer.py#L489-L497

Also, see: https://github.com/modular/modular/pull/5604/changes#r2685656362

KCaverly commented on max/python/max/interfaces/pipeline_variants/pixel_generation.py:
```diff
@@ -0,0 +1,315 @@
+    the available models on the server and determines the behavior and
+    capabilities of the response generation.
+    """
+    prompt: str | None = None
```
Is there any current functionality which does not require prompt to be provided? If so, we can remove the default of None and the optionality, and ensure that the prompt str is always provided, during init.

tolgacangoz replied:
I coded it as you mentioned here first, but then I realized that the text encoder of the pipeline processes with role and content style:

https://github.com/huggingface/diffusers/blob/9d68742214b9491718487b63f25d6e8236ed56ea/src/diffusers/pipelines/z_image/pipeline_z_image.py#L213-L240

I imagined the request like that, since this seemed to me to require the least modification at the `TextTokenizer`:
```console
curl -X POST http://localhost:8000/v1/images/generations \
  -H "Content-Type: application/json" \
  -d '{
    "messages": [{
                  "role": "user",
                  "content": "A cute baby sea otter floating in crystal clear water"
                  }],
    "chat_template_options": [{
                  "add_generation_prompt": True,
                  "enable_thinking": True
                  }],
    "n": 1,
    "size": "1024x1024",
    "guidance_scale": 0.0,
    "num_inference_steps": 9
  }' -o result.json
```

Or another alternative, we take the input as `prompt`, then wrap it to the role and content style `if isinstance(request, PixelGenerationRequest)`. But AFAIK, most diffusion models' text encoders don't use this role and content style input.
By the way, I am not familiar with this role and content style text processing.

KCaverly replied:
This is pretty common with Text models. What we commonly do if we want to support raw text prompts or messages. Is allow the user to pass both, and handle in the Tokenizer when producing the initial token ids. That seems pretty reasonable to do here. If a prompt is provided, and the tokenizer needs it wrapped as a one item message, we can handle that in the tokenizer itself. What we likely do not want to do, is introduce code upstream of the Request object, to wrap a single text prompt into a wrapped dict.

As a side note - It looks like the tokenizer is bypassed completely in this PR, and the model.py file, is stubbed out for conversion between text prompts and token ids. I would suggest you iron out working code for this, as a next step, this is an important step to identify for perf/arch etc.

tolgacangoz replied:
Exactly, I was imagining doing such a thing in the `TextTokenizer`. But saying `if isinstance(request, PixelGenerationRequest)` was naive since not all diffusion models need that wrapping.
How is this approach:

<details>
<summary><code>git diff</code></summary>

```diff
diff --git a/max/python/max/pipelines/architectures/z_image_module_v3/arch.py b/max/python/max/pipelines/architectures/z_image_module_v3/arch.py
index 707c1452b8..722aeb5651 100644
--- a/max/python/max/pipelines/architectures/z_image_module_v3/arch.py
+++ b/max/python/max/pipelines/architectures/z_image_module_v3/arch.py
@@ -11,6 +11,8 @@
 # limitations under the License.
 # ===----------------------------------------------------------------------=== #
 
+from functools import partial
+
 from max.graph.weights import WeightsFormat
 from max.interfaces import PipelineTask, PixelGenerationContext
 from max.nn.kv_cache import KVCacheStrategy
@@ -37,7 +39,14 @@ z_image_module_v3_arch = SupportedArchitecture(
     # scheduler=FlowMatchEulerDiscreteScheduler,
     # vae=AutoencoderKL,
     # text_encoder=Qwen3,
-    tokenizer=TextTokenizer,
+    tokenizer=partial(
+        TextTokenizer,
+        wrap_prompt_as_chat=True,
+        default_chat_template_options={
+            "add_generation_prompt": True,
+            "enable_thinking": True,
+        },
+    ),
     # transformer=ZImageTransformer2DModel,
     context_type=PixelGenerationContext,
     rope_type=RopeType.normal,
diff --git a/max/python/max/pipelines/lib/tokenizer.py b/max/python/max/pipelines/lib/tokenizer.py
index 438553c3b1..4998839cb1 100644
--- a/max/python/max/pipelines/lib/tokenizer.py
+++ b/max/python/max/pipelines/lib/tokenizer.py
@@ -239,6 +239,8 @@ class TextTokenizer(
         context_validators: list[Callable[[TextContext | PixelContext], None]]
         | None = None,
         subfolder: str | None = None,
+        wrap_prompt_as_chat: bool = False,
+        default_chat_template_options: dict[str, Any] | None = None,
         **unused_kwargs,
     ) -> None:
         self.model_path = model_path
@@ -298,6 +300,10 @@ class TextTokenizer(
             context_validators if context_validators else []
         )
 
+        # Prompt wrapping and default chat template options for diffusion models with chat-based text encoders
+        self._wrap_prompt_as_chat = wrap_prompt_as_chat
+        self._default_chat_template_options = default_chat_template_options
+
         if pipeline_config:
             huggingface_config = pipeline_config.model.huggingface_config
             if eos_token_id := getattr(
@@ -497,12 +503,28 @@ class TextTokenizer(
         self, request: TextGenerationRequest | PixelGenerationRequest
     ) -> TextContext | PixelContext:
         """Create a new TextContext or PixelContext object, leveraging necessary information from TextGenerationRequest or PixelGenerationRequest."""
+        # For diffusion models with chat-based text encoders, wrap raw prompts
+        # as user messages so they get properly chat-templated
+        prompt = request.prompt
+        messages = request.messages
+        chat_template_options = request.chat_template_options
+        if (
+            self._wrap_prompt_as_chat
+            and isinstance(request, PixelGenerationRequest)
+            and prompt is not None
+            and messages is None
+        ):
+            messages = [{"role": "user", "content": prompt}]
+            prompt = None
+            if self._default_chat_template_options:
+                chat_template_options = self._default_chat_template_options
+
         # Encode Prompt / Messages
         _prompt, token_ids = await self._generate_prompt_and_token_ids(
-            prompt=request.prompt,
-            messages=request.messages,
+            prompt=prompt,
+            messages=messages,
             tools=request.tools,
-            chat_template_options=request.chat_template_options,
+            chat_template_options=chat_template_options,
         )
 
         json_schema = (
```

</details>

I deactivated the text encoder and its tokenizer since it doesn't seem to work: https://github.com/modular/modular/issues/5708
Update: It now works with the latest nightly.

tolgacangoz replied:
I may have misunderstood your last paragraph. Could you elaborate on it?

KCaverly replied:
To elaborate on the ZImagesInputs comment. Code is here:

https://github.com/modular/modular/pull/5604/changes#diff-55418dc50e5cae03f70ed78b374dca4edaa306ba1bcc41f9b7b2b26ad915110eR109

I can't find the immediate code reference I was concerned about, but ZImagesInputs has a few items, I am unsure about:
1. It includes both text prompts & token ids. The goal of the Tokenizer & PixelGenerationContext is to produce request level inputs. I don't think we should be including text in this class at all, as text strings cannot be used directly by the model.
2. It is written from the perspective of a single request. Can we batch requests in the ZImageInputs pipeline, the expectation is that most of this class should be in Tensor space, for immediate execution in the forward pass of the model?
3. It does a ton of preprocessing for image inputs outside the graph. We will want to ensure as much as possible for this sits within the PixelTokenizer itself.

In general some best practices for this would include:
- Precompute as much data as possible inside the PixelTokenizer. If it is used in the model, and is unique to only the request, it should be produced in the PixelTokenizer by default, unless we can articulate a performance benefit for keeping it in the model.
- Keep the ZImageInputs object, mostly reserved to the direct runtime tensor params used in the pipeline itself. Note in most scenarios, constructing these objects, should be fairly lightweight. For example, in our most performant image models, this is little more than concatenating request level image data into batched tensors. 

I'll take a look through the tokenizer changes in this PR, and we can build out from there to the ZImageInputs object.

tolgacangoz replied:
Thank you for the clarification, Kyle!

1. You are definitely right. My priority was making the model functionally correct, meaning just producing the expected result. Now we are in the process of making this correct in terms of design, too. There might still have been relic parts.
2. I prioritized `batch_size=1` initially to complete the third stage -"Make it fast" with `batch_size=1`. I thought that we could have merged this model after the completion of that stage. When I study schedulers of the serving systems, at that point, I plan to add (continuous) batching strategies for throughput efficiency as the fourth stage -"Make it the fastest". I think we will need several strategies, such as arranging batches for different resolution requests, etc. I guess adding batching capability will be one of the latest PRs of the visual generation support series. WDYT?
3. I think you are refering all the calculations before the denoising loop. I think we should compile the whole `execute()`, right? Or should we move all the code that is before the denoising loop, including one-time operations like `text_encoder`'s generation of `prompt_embeds`, generation of the initial noise tensor of `latents`, etc., into the `PixelGenerationTokenizer/Processor`? I tried compiling the whole `execute()` too. You can examine the first comment of this PR for details. I am proceeding with moving all the code before the denoising loop into the `PixelGenerationTokenizer/Processor`. Actually, you are suggesting moving all the code even before the transformer's forward pass inside the denosing loop, right; that is, moving all the code except the transformer, scheduler's step, vae's decoding into `PixelGenerationTokenizer/Processor`? I need to be sure about these:
- Initial noise `Tensor`'s creation, aka. `prepare_latents()`,
- `Qwen3`'s initial one-time production of `prompt_embeds` by using the output of its tokenizer,
- Generation of timestep schedule from the scheduler, the number of inference steps, aka. `retrieve_timesteps()`, and other deterministic calculations inside the denoising loop on `timestep`.

I am trying to understand how strictly I should follow "Precompute as much data as possible". Would putting a compiled graph of `Qwen3`, the generation of the initial noise `Tensor` `latents`, and `timesteps` generation steps outside of the `model.py` be proper :thinking:? This seems weird/different to me, who is accustomed to HF's `diffusers`.

Am I supposed to design two large stages like "all the one-time preparation" and "the denoising loop + VAE's decoding"? In this case, where should we put `Qwen3`'s `prompt_embeds` (and `negative_prompt_embeds`) generation? It is a one-time preparation, but it needs a GPU; thus, should/can a "Tokenizer" use a GPU?
Also, should we name this as `PixelGenerationTokenizer`, or `PixelGenerationProcessor` ~to imply the addition of image/video conditioning too~ (actually, `TextAndVisionTokenizer` already has that feature despite being a "Tokenizer")? `PixelTokenizer` seems like tokenization of input pixels, right?

**AFAIU, the answers seem to be "Yes" and ~"`PixelGenerationProcessor`"~ `PixelGenerationTokenizer`, but I am not sure about the `Qwen3`'s proper place, though. WDYT?**

KCaverly replied:
> 2\. I prioritized `batch_size=1` initially to complete the third stage -"Make it fast" with `batch_size=1`. I thought that we could have merged this model after the completion of that stage. When I study schedulers of the serving systems, at that point, I plan to add (continuous) batching strategies for throughput efficiency as the fourth stage -"Make it the fastest". I think we will need several strategies, such as arranging batches for different resolution requests, etc. I guess adding batching capability will be one of the latest PRs of the visual generation support series. WDYT?

I think this is correct. I think its fine to prioritize a batch size of 1 for model bringup and accuracy. We can move from bs=1+ once everything else is correct. That said, we should build the interfaces with batches in mind. As far as `ModelInputs` are concerned these should be created assuming we move to a batch size > 1, and no inputs should be provided, which are specific to a single request. All of these inputs should represent batched inputs.

Let me take a look through the second point for the tokenizer questions.

KCaverly commented on max/python/max/interfaces/pipeline_variants/pixel_generation.py:
```diff
@@ -0,0 +1,315 @@
+            raise RuntimeError("Either prompt or messages must be provided.")
+
+
+class PixelGenerationMetadata(
```
Can we remove this class, and any of its use in this PR. It does not appear to be populated or used anywhere.

tolgacangoz replied:
I took inspiration from `AudioGenerationMetadata`:
https://github.com/modular/modular/blob/38c1a3745c5fdffacc816ae14a6d47bef8660f40/max/python/max/interfaces/pipeline_variants/audio_generation.py#L80-L88
Other than test files, the two are seen in 7 places in the repo. Should the two be removed?

I should imitate `TextGenerationOutput`, right, since it doesn't have any `metadata` part like `AudioGenerationOutput`:
https://github.com/modular/modular/blob/38c1a3745c5fdffacc816ae14a6d47bef8660f40/max/python/max/interfaces/pipeline_variants/audio_generation.py#L176-L179

KCaverly replied:
Yeah, I would use the TextGeneration... objects as a proper best practice. This is on use to make this clearer, but Audio has a bunch of tech debt, we have yet to clean up.

tolgacangoz replied:
Alright, I got it.

KCaverly commented on max/python/max/interfaces/pipeline_variants/pixel_generation.py:
```diff
@@ -0,0 +1,315 @@
+
+
+@runtime_checkable
+class PixelGenerationContext(BaseContext, Protocol):
```
Please make the same changes to the Request Above. Remove any fields which are not yet used in the downstream pipeline.

tolgacangoz replied:
Done!

KCaverly commented on max/python/max/interfaces/pipeline_variants/pixel_generation.py:
```diff
@@ -0,0 +1,315 @@
+    request IDs to context objects.
+    """
+
+    num_steps: int = 1
```
Please remove this attribute it does not appear to be used downstream.

tolgacangoz replied:
Right, removed.

KCaverly commented on max/python/max/interfaces/pipeline_variants/pixel_generation.py:
```diff
@@ -0,0 +1,315 @@
+        default_factory=lambda: np.array([], dtype=np.float32)
+    )
+    """The generated pixel data (image/video frames), if available."""
+    metadata: PixelGenerationMetadata = msgspec.field(
```
Please remove this, as it is never constructed with non-default values downstream, and does not appear to be used.

tolgacangoz replied:
Removed!

Also, see: https://github.com/modular/modular/pull/5604/changes#r2685991136

KCaverly commented on max/python/max/interfaces/pipeline_variants/pixel_generation.py:
```diff
@@ -0,0 +1,315 @@
+
+    final_status: GenerationStatus
+    """The final status of the generation process."""
+    steps_executed: int
```
How is steps_executed used, is this returned to the user?

tolgacangoz replied:
As a second thought, this is not important for now, since Z-Image-Turbo is fast enough, so there is no need for something like [displaying intermediate denoising stages](https://huggingface.co/docs/diffusers/main/en/using-diffusers/callback#display-intermediate-images). Removed for now.

KCaverly reviewed (commented):
This is a great start! To start us off, we will need to break up this PR a little to make it easier to digest and integrate it into our code base. 

I think the best place to start is the changes to max/interfaces. I've added a series of comments to this folder, on thoughts we will want to look at. It looks like you've added a bunch of forward looking functionality (which is great), but will make this more challenging, as it introduces code into the code base, which is not (yet) actively used. 

I would suggest, that you take a look at some of the comments suggested in the max/interfaces folder, and create a new PR with just those changes. This is a pretty core piece of our infra, and will want to land first, so that other image generation pipelines, can start to use this functionality as well.

Once those land those we can build out larger and larger building blocks for integrating the entire thing into the code base.
